src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/UnnecessaryParenthesesCheck.java

```diff
@@ -607,21 +660,85 @@ else if (assignDepth >= 1) {
      *         unnecessary parentheses.
      */
     private static boolean unnecessaryParenAroundOperators(DetailAST ast) {
+        final int type = ast.getType();
+        final boolean isConditionalOrRelational = TokenUtil.isOfType(type, CONDITIONAL_OPERATOR)
+                        || TokenUtil.isOfType(type, RELATIONAL_OPERATOR);
+        final boolean isBitwise = TokenUtil.isOfType(type, BITWISE_BINARY_OPERATORS);
+        final boolean hasUnnecessaryParentheses;
+        if (isConditionalOrRelational) {
+            hasUnnecessaryParentheses = checkConditionalOrRelationalOperator(ast);
+        }
+        else if (isBitwise) {
+            hasUnnecessaryParentheses = checkBitwiseBinaryOperator(ast);
+        }
+        else {
+            hasUnnecessaryParentheses = TokenUtil.isOfType(type, UNARY_AND_POSTFIX)
+                    && isBitWiseBinaryOrConditionalOrRelationalOperator(ast.getParent().getType());
+        }
+        return hasUnnecessaryParentheses;
+    }
+
+    /**
+     * Check if conditional or logical operator has unnecessary parentheses.
```

Comment:
Shouldn't logical be relational? The method is named `OrRelational` but your javadoc says logical.

Reply:
done